src/tool_parsecfg.c

```diff
@@ -335,39 +346,22 @@ static const char *unslashquote(const char *line, char *param)
 
 /*
  * Reads a line from the given file, ensuring is NUL terminated.
- * The pointer must be freed by the caller.
- * NULL is returned on an out of memory condition.
  */
-static char *my_get_line(FILE *fp)
+static bool my_get_line(FILE *fp, struct curlx_dynbuf *db,
+                        bool *error)
 {
   char buf[4096];
-  char *nl = NULL;
-  char *line = NULL;
-
+  *error = FALSE;
   do {
-    if(NULL == fgets(buf, sizeof(buf), fp))
-      break;
-    if(!line) {
-      line = strdup(buf);
-      if(!line)
-        return NULL;
+    /* fgets() returns s on success, and NULL on error or when end of file
+       occurs while no characters have been read. */
+    if(!fgets(buf, sizeof(buf), fp))
+      return FALSE; /* stop reading */
```

Comment:
We have to return true if there was data written to the buffer before eof and return false if there was no data written to the buffer before eof, that is why we need the check I suggested. The function in master has similar behavior, it returns NULL if there's no line or a pointer to the line if there's a line, even if feof.

If you want to handle error on data failed to write to the buffer then ideally I think we'd reset the buffer to its previous state and return false. Having a separate `error` just complicates things IMO. (I'm not sure I follow the logic but I just skimmed it so maybe it is right.) Currently there's no way to clip with dyn buffer but I will write a dyn function to do it like std:;string erase, which I think may be useful anyway.

Reply:
> We have to return true if there was data written to the buffer before eof

No, not if `*_dyn_add()` fails as then it'll be partial.

> The function in master has similar behavior, it returns NULL if there's no line or a pointer to the line if there's a line, even if feof.

This version returns all data until EOF, unless fgets() returns an error prematurely (I'm not sure what kind of condition that could be) or dyn_add fails.

The original version didn't bail out on memory error, otherwise it is a similar behavior.

> If you want to handle error on data failed to write to the buffer then ideally I think we'd reset the buffer to its previous state and return false.

I disagree. If dyn_add fails, it clears the entire buffer, we return error and won't parse any more.

> Having a separate error just complicates things IMO

It makes a memory problem fail the parsing instead of just truncating the parsing.

Reply:
I'm saying this:

printf "a" > file
char buf[2];
fgets(buf, sizeof(buf), fp) <-- returns buf and no eof
loop because no \n
fgets(buf, sizeof(buf), fp) <-- returns NULL and eof

We should not return false in this case, the line is "a". Concerning error if we have only a piece of a line already added and an error occurs before adding the next piece then I would remove what was added since it's incomplete. If you are only having a single line in the buffer at any time then that is not necessary you could just reset it. I've submitted #5958 which will allow truncating a dyn buffer which I think is a good idea to have anyway, in case we need it at some point.